In [1]:
import bs4 as bs
import urllib.request
import pandas as pd

In [ ]:
# Reading the html file
source = urllib.request.urlopen('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').read()

# Finding the table
soup = bs.BeautifulSoup(source, 'html.parser')
table = soup.find('table', attrs={'class':'wikitable sortable'})

# Get all table rows
table_rows = table.find_all('tr')

# Taking the column names of the table
ths = table_rows[0].find_all('th')
cols =[th.text.strip().replace(" ", "") for th in ths]

# Extracting all the data in the table
data = [] 
table_rows[0]
for tr in table_rows:
    tds = tr.find_all('td')
    row = [td.text.strip() for td in tds]
    if len(row) > 0:
        data.append(row)

In [ ]:
# Crerating the dataframe
df_table = pd.DataFrame(data, columns=cols)
df_table.head()

Checking for repeated postal codes

In [ ]:
df_table[['PostalCode','Borough']].groupby('PostalCode').count().sort_values(by='Borough',ascending=False)

In [ ]:
df_table[df_table['PostalCode'] == 'M5A']

Therefore, in the the new table in wikipedia, postal codes are not repeated. Now lets removed Boroughs that are Not assigned.

In [6]:
df_table = df_table[df_table['Borough'] != 'Not assigned']

Now not assigned Boroughs are removed from the table. Now lets check if we have any Neighbourhoods with Not Assigned value.

In [9]:
df_table[df_table['Neighbourhood'] != 'Not assigned']

In [ ]:
This shows that there are there are Neighbourhoods with Not Assigned values. Therefore we are done cleaning the table.

In [7]:
df_table.shape

(103, 3)

Cleaned table have 103 rows that corresponds to 103 Boroughs.